include/mbgl/style/expression/compound_expression.hpp

```diff
@@ -0,0 +1,334 @@
+#pragma once
+
+#include <array>
+#include <vector>
+#include <memory>
+#include <mbgl/util/optional.hpp>
+#include <mbgl/util/variant.hpp>
+#include <mbgl/util/color.hpp>
+#include <mbgl/style/expression/expression.hpp>
+#include <mbgl/style/expression/type.hpp>
+#include <mbgl/style/expression/value.hpp>
+#include <mbgl/style/expression/parsing_context.hpp>
+#include <mbgl/style/conversion.hpp>
+
+#define expand_pack(...) (void) std::initializer_list<int>{((void)(__VA_ARGS__), 0)...};
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+namespace detail {
+    template <typename T>
+    std::decay_t<T> get(const Value& value);
+} // namespace detail
+
+
+struct VarargsType { type::Type type; };
+template <typename T>
+struct Varargs : std::vector<T> { using std::vector<T>::vector; };
+
+
+struct SignatureBase {
+    SignatureBase(type::Type result_, variant<std::vector<type::Type>, VarargsType> params_) :
+        result(result_),
+        params(params_)
+    {}
+    virtual ~SignatureBase() {}
+    virtual std::unique_ptr<TypedExpression> makeTypedExpression(std::vector<std::unique_ptr<TypedExpression>>) const = 0;
+    type::Type result;
+    variant<std::vector<type::Type>, VarargsType> params;
+};
+
+template <class, class Enable = void>
+struct Signature;
+
+// Signature from a zoom- or property-dependent evaluation function:
+// (const EvaluationParameters&, T1, T2, ...) => Result<U>,
+// where T1, T2, etc. are the types of successfully-evaluated subexpressions.
+template <class R, class... Params>
+struct Signature<R (const EvaluationParameters&, Params...)> : SignatureBase {
```

Comment:
`Signature` is probably a misnomer, since this object holds not just the types, but also a pointer to an expression's `evaluate` implementation.